RFC-0012-thrift-serialization-for-task-updates.md

```diff
@@ -0,0 +1,394 @@
+        jsonBinder(binder).addModuleBinding().to(SplitJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(OutputTableHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(InsertTableHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(DeleteTableHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(IndexHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(TransactionHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(PartitioningHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(FunctionHandleJacksonModule.class);
+        jsonBinder(binder).addModuleBinding().to(MetadataUpdateJacksonModule.class);
+
+        binder.bind(HandleResolver.class).in(Scopes.SINGLETON);
+    }
+}
+
+// A handle resolver to return the correct type info in runtime
+public HandleResolver()
+{
+    handleResolvers.put(REMOTE_CONNECTOR_ID.toString(), new MaterializedHandleResolver(new RemoteHandleResolver()));
+    handleResolvers.put("$system", new MaterializedHandleResolver(new SystemHandleResolver()));
+    handleResolvers.put("$info_schema", new MaterializedHandleResolver(new InformationSchemaHandleResolver()));
+    handleResolvers.put("$empty", new MaterializedHandleResolver(new EmptySplitHandleResolver()));
+
+    functionHandleResolvers.put("$static", new MaterializedFunctionHandleResolver(new BuiltInFunctionNamespaceHandleResolver()));
+    functionHandleResolvers.put("$session", new MaterializedFunctionHandleResolver(new SessionFunctionHandleResolver()));
+}
+
+// Register correct serde methods for different types
+protected AbstractTypedJacksonModule(
+            Class<T> baseClass,
+            Function<T, String> nameResolver,
+            Function<String, Class<? extends T>> classResolver)
+{
+    super(baseClass.getSimpleName() + "Module", Version.unknownVersion());
+
+    TypeIdResolver typeResolver = new InternalTypeResolver<>(nameResolver, classResolver);
+
+    addSerializer(baseClass, new InternalTypeSerializer<>(baseClass, typeResolver));
+    addDeserializer(baseClass, new InternalTypeDeserializer<>(baseClass, typeResolver));
+}
+
+// type information
+private static final String TYPE_PROPERTY = "@type";
+
+public InternalTypeSerializer(Class<T> baseClass, TypeIdResolver typeIdResolver)
+{
+    super(baseClass);
+    this.typeSerializer = new AsPropertyTypeSerializer(typeIdResolver, null, TYPE_PROPERTY);
+}
+
+public InternalTypeDeserializer(Class<T> baseClass, TypeIdResolver typeIdResolver)
+{
+    super(baseClass);
+    this.typeDeserializer = new AsPropertyTypeDeserializer(
+            TypeFactory.defaultInstance().constructType(baseClass),
+            typeIdResolver,
+            TYPE_PROPERTY,
+            false,
+            null);
+}
+
+```
+
+### Preferred Option:  Pluggable Serde for Polymorphic Types
```

Comment:
@shangm2 : Would be great to see an example of what the native code corresponding to this would look like. Can you add some code-sample for that ?